src/webgl-surface/shaders/uniform-ribbon.vs

```diff
@@ -0,0 +1,156 @@
+
+  vec2 c1 = block3.xy;
+  vec2 c2 = block3.zw;
+
+  float depth = block4.x;
+  float resolution = block4.y;
+  vec2 threshold = block4.zw;
+
+  float normalDirection = position.x;
+  float vertexIndex = position.y;
+  float instance = position.z;
+
+
+  vec2 currentPosition;
+
+  vertexColor = mix(pickColor(startColor), pickColor(endColor), vertexIndex / resolution);
+
+  if (vertexIndex < threshold.x ) {
+    float realTime = vertexIndex / threshold.x;
+
+    // radius of hemiSphere where endpoints are in
+    float r1 = distance(c1, start1);
+
+    // mid of two end points
+    vec2 mid1 = getMiddle(start1, start2);
+
+    // distance from center to getMiddle;
+    float d1 = distance(mid1, c1);
+
+    // radius - distance
+    float l1 = r1 - d1;
+
+    // angle between line(middle, center) and line(currentPosition, center)
+    float cosRadian = acos((r1 - l1 * realTime) / r1);
+
+    // rotation angle of line(middle, center)
+    float radian = acos((mid1.x - c1.x) / d1);
+
+    if ((mid1.y - c1.y) < 0.0) radian = - radian;
+
+    if (normalDirection == 1.0) {
+      currentPosition = c1 + r1 * vec2(cos(radian - cosRadian), sin(radian - cosRadian));
+    }
+
+    else if (normalDirection == -1.0) {
+      currentPosition = c1 + r1 * vec2(cos(radian + cosRadian), sin(radian + cosRadian));
+    }
+  }
+
+  else if (vertexIndex >= threshold.x  && vertexIndex <= resolution - threshold.y ) {
+    float realTime = (vertexIndex - threshold.x) / (resolution - threshold.x - threshold.y);
+
+    if (normalDirection == 1.0) {
+      currentPosition = makeBezier2(realTime, start1, end1, controlPoint);
+    }
+
+    else if (normalDirection == -1.0) {
+      currentPosition = makeBezier2(realTime, start2, end2, controlPoint);
+    }
+
+    float minMidFade = 0.76;
+    float stop1 = 0.17;
+    float stop2 = 0.83;
+
+    if (realTime > stop2) {
+      vertexColor *= mix(minMidFade, 1.0, (realTime - stop2) / stop1);
+    }
+
+    else if (realTime < stop1) {
+      vertexColor *= mix(minMidFade, 1.0, (stop1 - realTime) / stop1);
+    }
+
+    else {
+      vertexColor *= minMidFade;
+    }
+  }
+
+  else if (vertexIndex > resolution - threshold.y ) {
+    float realTime = (vertexIndex - resolution + threshold.y) / threshold.y;
+
+    float r2 = distance(c2, end1);
+
+    vec2 mid2 = getMiddle(end1, end2);
+
+    float d2 = distance(mid2, c2);
+
+    float l2 = r2 - d2;
+
+    float cosRadian = acos((d2 + l2 * realTime) / r2);
+
+    float radian = acos((mid2.x - c2.x) / d2);
+
+    if ((mid2.y - c2.y) < 0.0) radian = - radian;
+
+    if (normalDirection == 1.0) {
+      currentPosition = c2 + r2 * vec2(cos(radian + cosRadian), sin(radian + cosRadian));
+    }
+
+    else if (normalDirection == -1.0) {
+      currentPosition = c2 + r2 * vec2(cos(radian - cosRadian), sin(radian - cosRadian));
+    }
+  }
+
+  vec4 mvPosition = modelViewMatrix * vec4(currentPosition, position.z, 1.0);
+  gl_Position = projectionMatrix * mvPosition;
+}
```

Comment:
don't put newlines between each statement unless it's a code block or you have a comment.